# HG changeset patch
# User Rik <rik@octave.org>
# Date 1533096188 25200
#      Tue Jul 31 21:03:08 2018 -0700
# Node ID 02fbbde153082d0187373586585f619a9e320447
# Parent  ca38400776bc96f6b48abf7963c6ec94024aaac3
Fix problems with cset 39fd627b2c5d when octave_idx_type is 64-bit (bug #54405).

* oct-stream.cc (get_size): Avoid using abs() on int64 data types as it returns
incorrect values.  Instead, check that numbers are positive before comparing.

diff --git a/libinterp/corefcn/oct-stream.cc b/libinterp/corefcn/oct-stream.cc
--- a/libinterp/corefcn/oct-stream.cc
+++ b/libinterp/corefcn/oct-stream.cc
@@ -180,19 +180,19 @@ namespace octave
 
     nr = get_size (dnr, who);
 
     if (dnc >= 0.0)
       {
         nc = get_size (dnc, who);
 
         // Check for overflow.
-        if (nr != 0 &&
-            abs (nc) > abs (std::numeric_limits<octave_idx_type>::max () / nr))
-         ::error ("%s: size too large for Octave's index type", who.c_str ());
+        if (nr > 0 && nc > 0
+            && nc > std::numeric_limits<octave_idx_type>::max () / nr)
+          ::error ("%s: size too large for Octave's index type", who.c_str ());
       }
   }
 
   static std::string
   expand_char_class (const std::string& s)
   {
     std::string retval;
 
